Holistic Evaluation of Text-to-Image Models (HEIM)

teetone opened the pull request:
Merging [HEIM](https://github.com/stanford-crfm/heim) changes (scenarios, models, and metrics) changes into HELM.

yifanmai commented on src/helm/benchmark/metrics/image_generation/detectors/vitdet.py:
```diff
@@ -0,0 +1,171 @@
+
+class ViTDetDetector(BaseDetector):
+    def __init__(self):
+        from detectron2.checkpoint import DetectionCheckpointer
+        from detectron2.config import LazyConfig
+        from detectron2.config import instantiate
+        from detectron2.data.catalog import MetadataCatalog
```
For all optional dependencies, could you do:

```python
try:
    from detectron2.checkpoint import DetectionCheckpointer
    from detectron2.config import LazyConfig
    from detectron2.config import instantiate
    from detectron2.data.catalog import MetadataCatalog
except ModuleNotFoundError as e:
    handle_module_not_found_error(e, ["heim"])
```

The `["heim"]` part requires #1944 to be merged first though.

Here's an [example](https://github.com/stanford-crfm/helm/blob/50e656526c9946ad3bc2cf32be5a52656ac7deb6/src/helm/benchmark/metrics/summarization_metrics.py#L57-L60).

yifanmai reviewed (changes requested):
Looks promising, much better than the last PR!

Partial review, since 14k loc is too much for me to review at once.

Also, can we arrange a mini design review meeting for the schema changes?

yifanmai commented on setup.cfg:
```diff
@@ -126,6 +126,105 @@ images =
+    Werkzeug~=2.2.2
+    wrapt~=1.14.1
+
+deep_floyd =
+    accelerate~=0.19.0
+    dacite~=1.6.0
+    diffusers[torch]~=0.16.1
+    pyhocon~=0.3.59
+    pymongo~=4.2.0
+    retrying~=1.3.3
+    safetensors~=0.3.1
+    sentencepiece~=0.1.97
+    sqlitedict~=1.7.0
+    tqdm~=4.64.1
+    transformers~=4.29.2
+    zstandard~=0.18.0
+
```
Could you move this to its own `requirements-deep-floyd.txt` file? `setup.cfg` should be just for HELM stuff.

teetone replied:
I will move these to the deepfloyd script.

yifanmai commented on setup.cfg:
```diff
@@ -126,6 +126,105 @@ images =
     accelerate~=0.23.0  # For the newer versions of Transformers
     pillow~=9.4.0
 
+heim =
```
Are all of these needed?

In general, this list should only be _direct_ dependencies, not _transitive_ dependencies.

teetone commented on setup.cfg:
```diff
@@ -164,7 +264,10 @@ exclude =
-exclude = venv/*
+exclude =
+    venv/*
+    src/helm/proxy/clients/image_generation/dalle_mini/*
+    src/helm/proxy/clients/image_generation/mindalle/*
```
We need these files to run local inference for these two models.

teetone commented on src/helm/benchmark/augmentations/translate_perturbation.py:
```diff
@@ -0,0 +1,29 @@
+
+    @dataclass(frozen=True)
+    class Description(PerturbationDescription):
+        language_code: str = "zh-CN"
```
We need this to differentiate instances by the language code when computing metrics. This is similar to what we did for `prob` in `TyposPerturbation`.

teetone commented on src/helm/common/request.py:
```diff
@@ -120,6 +137,9 @@ class Sequence:
     # Why did the sequence finish?
     finish_reason: Optional[Dict] = None
 
+    # Could be a sequence made up of multimedia content
+    multimodal_content: Optional[MultimediaObject] = None
```
We are going to rename `Sequence` to `Completion`. Will follow up on this in a separate PR.

teetone commented on src/helm/benchmark/scenarios/scenario.py:
```diff
@@ -62,6 +62,9 @@ class Input:
     """A single input can consists of multimodal content interleaved (e.g., text, image, text, ...)."""
 
+    original_text: Optional[str] = None
+    """Unperturbed text before any perturbations are applied."""
```
I believe that's only possible if `should_include_original_eval` of `DataAugmenterSpec` is `True`, but by default, it's `False`: https://github.com/stanford-crfm/helm/blob/main/src/helm/benchmark/augmentations/data_augmenter.py#L91. 

Even if `should_include_original_eval` is set to `True`, it's inconvenient to find the original instance for a particular perturbed instance and get the original text on the metric level because we usually compute metrics instance by instance (most metrics implement `evaluate_generation`, for example). So, instead of `original_text`, I added a new field `original_instance: Optional[Instance]` to `Instance`. This solution is general enough that it should work for the multimodal case (e.g., perturbed vs. original images). What do you think about this?

teetone commented:
TODO: fixing merge conflicts from https://github.com/stanford-crfm/helm/pull/1903

yifanmai commented:
Please address the merge conflict in `src/helm/proxy/clients/auto_client.py`.

Please also address test failures:
```
=========================== short test summary info ============================
FAILED src/helm/benchmark/test_model_properties.py::TestModelProperties::test_num_models_available - AssertionError: assert 120 == 119
 +  where 120 = len([ModelDeployment(name='simple/model1', client_spec=ClientSpec(class_name='helm.proxy.clients.simple_client.SimpleClien...x_sequence_length=6000, max_request_length=None, max_sequence_and_generated_tokens_length=None, deprecated=False), ...])
FAILED src/helm/benchmark/adaptation/adapters/test_image_generation_adapter.py::TestImageGenerationAdapter::test_construct_prompt - ValueError: Model deployment  not found
ERROR src/helm/benchmark/window_services/image_generation/test_clip_window_service.py::TestCLIPWindowService::test_truncate_from_right - ValueError: Model deployment huggingface/dreamlike-photoreal-v2-0 not found
ERROR src/helm/benchmark/window_services/image_generation/test_dalle2_window_service.py::TestDALLE2WindowService::test_fits_within_context_window - ValueError: Model deployment openai/dalle-2 not found
ERROR src/helm/benchmark/window_services/image_generation/test_dalle2_window_service.py::TestDALLE2WindowService::test_truncate_from_right - ValueError: Model deployment openai/dalle-2 not found
= 2 failed, 424 passed, 11 skipped, 1 deselected, 2037 warnings, 3 errors in 244.35s (0:04:04) =
```

teetone commented:
> Please address the merge conflict in `src/helm/proxy/clients/auto_client.py`.
> 
> Please also address test failures:
> 
> ```
> =========================== short test summary info ============================
> FAILED src/helm/benchmark/test_model_properties.py::TestModelProperties::test_num_models_available - AssertionError: assert 120 == 119
>  +  where 120 = len([ModelDeployment(name='simple/model1', client_spec=ClientSpec(class_name='helm.proxy.clients.simple_client.SimpleClien...x_sequence_length=6000, max_request_length=None, max_sequence_and_generated_tokens_length=None, deprecated=False), ...])
> FAILED src/helm/benchmark/adaptation/adapters/test_image_generation_adapter.py::TestImageGenerationAdapter::test_construct_prompt - ValueError: Model deployment  not found
> ERROR src/helm/benchmark/window_services/image_generation/test_clip_window_service.py::TestCLIPWindowService::test_truncate_from_right - ValueError: Model deployment huggingface/dreamlike-photoreal-v2-0 not found
> ERROR src/helm/benchmark/window_services/image_generation/test_dalle2_window_service.py::TestDALLE2WindowService::test_fits_within_context_window - ValueError: Model deployment openai/dalle-2 not found
> ERROR src/helm/benchmark/window_services/image_generation/test_dalle2_window_service.py::TestDALLE2WindowService::test_truncate_from_right - ValueError: Model deployment openai/dalle-2 not found
> = 2 failed, 424 passed, 11 skipped, 1 deselected, 2037 warnings, 3 errors in 244.35s (0:04:04) =
> ```

@yifanmai I resolved the merge conflicts, fixed the tests, and added all the text-to-image model deployments and metadata. Could you take another look when you get the chance? Thanks!

yifanmai commented on src/helm/benchmark/scenarios/scenario.py:
```diff
@@ -148,6 +148,9 @@ class Instance:
     contrast_references: Optional[List[List[Reference]]] = None
     """References for the perturbed input above (if available)"""
 
+    original_instance: Optional["Instance"] = None
```
This means that original instances are stored twice - once in the outermost instances list, and once as a nested `original_instance` in the perturbed instance.

Is there a design that does not do this double storing? For instance, in the metrics, you can access the original instances by looking for the corresponding instance with the same ID and without a perturbation description.

teetone replied:
@yifanmai, What do you mean by the outermost instances list? I don't think we store the original instances unless `should_include_original_eval` is set to True and that value is not always True (see this comment https://github.com/stanford-crfm/helm/pull/1939#discussion_r1392160329).

yifanmai reviewed (changes requested):
Partial review.

teetone commented on install-heim-extras.sh:
```diff
@@ -0,0 +1,39 @@
+else
+    ROOT=`exec pwd`
+    mkdir -p tmp && chmod -R 777 tmp && rm -r tmp
+    mkdir -p tmp && cd tmp && git clone https://github.com/michiyasunaga/apex && cd apex && git checkout 9395ba2aab3c05e0e36ef0b7fe48d42de9f10bcf && pip install -v --disable-pip-version-check --no-cache-dir ./
```
The arguments between the two apex installations (inside the if block and inside the else block) are different

teetone commented on install-heim-extras.sh:
Fixed.

yifanmai commented on src/helm/benchmark/adaptation/adapter_spec.py:
```diff
@@ -71,6 +73,9 @@ class AdapterSpec:
     # set of training instances.  Used to compute error bars.
     num_train_trials: int = 1
 
+    # Number of trials, where we query the model with the exact same requests
```
Clarify that a different seed is used for each request - so it is not the "exact same" request.

yifanmai commented on src/helm/benchmark/augmentations/test_perturbation.py:
```diff
@@ -234,6 +235,17 @@ def test_gender_term_perturbation():
+    instance: Instance = Instance(id="id0", input=Input(text="A blue dog"), references=[])
+    instances: List[Instance] = data_augmenter.generate([instance], include_original=True)
+
+    print(instances)
```
Remove print.

yifanmai commented on src/helm/benchmark/augmentations/translate_perturbation.py:
```diff
@@ -0,0 +1,29 @@
+
+    @dataclass(frozen=True)
+    class Description(PerturbationDescription):
+        language_code: str = "zh-CN"
```
Remove the default value for `language_code` (not sure why "zh-CN" should be the default).

yifanmai commented on src/helm/benchmark/presentation/run_display.py:
```diff
@@ -43,6 +45,9 @@ class DisplayPrediction:
     truncated_predicted_text: Optional[str]
     """The truncated prediction text, if truncation is required by the Adapter method."""
 
+    base64_images: List[str]
```
This should be `Optional[List[str]]` for reverse compatibility.

yifanmai commented on src/helm/benchmark/metrics/image_generation/image_metrics_utils.py:
I'd suggest moving this to `helm/common/request.py` or a new file `helm/common/multimodal_request_utils.py` because this is used by non-metrics things (e.g. `run_display.py`)

yifanmai commented on src/helm/benchmark/run_expander.py:
```diff
@@ -435,6 +441,19 @@ class NumOutputsRunExpander(ReplaceValueRunExpander):
+    """For getting different generations for the same requests."""
+
+    name = "num_trials"
+    values_dict = {
```
File an issue / add a comment to clean this up?

teetone replied:
I removed the numbers and added `heim_efficiency`.

yifanmai commented on src/helm/benchmark/scenarios/test_math_scenario.py:
```diff
@@ -1,9 +1,14 @@
 
 
+@pytest.mark.skip(
+    reason="Incompatible with newer versions with diffusers>0.24.0. Fails with "
```
Does this mean the MATH scenario itself also does not work if diffusers is installed? If so, please open an issue / add a TODO comment to fix this.

yifanmai commented on src/helm/benchmark/test_model_properties.py:
You can optionally revert this file - the agreement I have with Joss is that new models will not be added to the constants lists, and we will delete the constants list eventually once we're done with all the WindowService migrations that we want to do. But also no harm leaving these changes in.

teetone replied:
I will leave these for now since these tests did catch some errors before, and we can remove them later!

yifanmai commented on src/helm/common/file_caches/local_file_cache.py:
```diff
@@ -0,0 +1,37 @@
+            return os.path.join(self._location, file_name)
+
+        file_path: str
+        while True:
+            file_path = generate_one()
+            if not os.path.exists(file_path):
+                break
```
Don't need to do this checking - it is incorrect anyway because there can be race conditions.

yifanmai commented on src/helm/common/file_caches/file_cache.py:
```diff
@@ -0,0 +1,21 @@
+        pass
+
+    @abstractmethod
+    def get_unique_file_location(self) -> str:
```
Rename to `generate_unique_new_file_path()` - it's not really a getter.

yifanmai replied:
Delete from the abstract class since this isn't use anywhere outside `LocalFileCache`.

yifanmai commented on src/helm/proxy/services/remote_service.py:
```diff
@@ -27,6 +31,8 @@ class RemoteServiceError(Exception):
 
 
 class RemoteService(Service):
+    NOT_SUPPORTED_ERROR: str = "Not supported through the remote service."
```
No action needed: We should eventually split all these miscellaneous things out of `Service`.

yifanmai reviewed (changes requested):
Partial review, will do the rest in a couple of hours.

yifanmai commented on src/helm/config/model_metadata.yaml:
```diff
@@ -557,6 +623,161 @@ models:
     tags: [VISION_LANGUAGE_MODEL_TAG]
 
+  ## Text-to-Image Diffusion Models
+  - name: huggingface/dreamlike-diffusion-v1-0
```
Oops, this is stale, resolving.

yifanmai commented on src/helm/config/model_deployments.yaml:
```diff
@@ -415,7 +497,172 @@ model_deployments:
+  ## Text-to-Image Diffusion Models
 
+  - name: huggingface/dreamlike-diffusion-v1-0
+    model_name: huggingface/dreamlike-diffusion-v1-0
```
Oops, this is stale, resolving.

yifanmai commented on src/helm/common/general.py:
```diff
@@ -20,6 +20,17 @@
 _CREDENTIALS_FILE_NAME = "credentials.conf"
 _CREDENTIALS_ENV_NAME = "HELM_CREDENTIALS"
 
+HELM_LOCAL_CACHE_PATH: str = ".helm_cache"
```
BTW there's now a way to get the benchmark_output path: https://github.com/stanford-crfm/helm/blob/a9e102a479afe46137c29a7fa72f3ce72d8a6eed/src/helm/benchmark/runner.py#L46

yifanmai commented on src/helm/benchmark/window_services/image_generation/lexica_search_window_service.py:
```diff
@@ -0,0 +1,20 @@
+from helm.benchmark.window_services.tokenizer_service import TokenizerService
+
+
+class LexicaSearchWindowService(CLIPWindowService):
```
Does this window service (and other window services in this pull request) have the same `truncate_from_right()` quirk as CLIPWindowService?

teetone replied:
No, so I overrode the parent method.

yifanmai commented on src/helm/common/file_caches/local_file_cache.py:
```diff
@@ -0,0 +1,37 @@
+
+        return file_path
+
+    def get_unique_file_location(self) -> str:
```
Rename to `generate_unique_new_file_path()` - it's not really a getter.

yifanmai commented on src/helm/proxy/clients/auto_client.py:
```diff
@@ -66,9 +68,17 @@ def _get_client(self, model_deployment_name: str) -> Client:
             host_organization: str = model_deployment.host_organization
             cache_config: CacheConfig = build_cache_config(self.cache_path, self.mongo_uri, host_organization)
 
+            # Initialize `FileCache` for text-to-image model APIs
+            local_file_cache_path: str = os.path.join(self.cache_path, "output", host_organization)
+            file_cache: FileCache = LocalFileCache(local_file_cache_path, file_extension="png")
```
Make `file_cache` a `provider_binding` rather than a `constant_binding` i.e. lazily create the folder in a lambda.

teetone replied:
Do you mind explaining `provider_binding` vs `constant_binding` a bit more and sharing how the lambda looks like?

yifanmai replied:
For `provider_binding`, the value should be a parameter-less function that returns the value, rather than the value itself. It can either be a lambda or a named method.

yifanmai replied:
I'll send you a PR for this.

yifanmai replied:
Opened PR #2175 for this.

yifanmai commented on src/helm/common/file_caches/local_file_cache.py:
```diff
@@ -0,0 +1,37 @@
+        with open(file_path, "wb") as f:
+            f.write(compute())
+
+        return file_path
```
I think this should be relative to `base_path`, as opposed to relative to the current working directory? Currently this means that I can't change my cache from "prod_env/cache" to anything else without also having to migrate all my requests strings.

yifanmai replied:
In this pull request, the `MediaObject` in `RequestResponse` in the `ScenarioState` requires a path. The HEIM pull requests populates this path with things like `prod_env/cache/openai/77437ea482144bf7b9275a0acee997db.png`. I have a bunch of issues with this:

1. The name file cache, and the location of the cache in `prod_env/cache`, suggests that it is temporary storage. However, it is not actually temporary - `prod_env/cache` contains the only copy of generated images, and deleting `prod_env/cache` results in loss of data that comprises part of the `scenario_state.json` files.
2. The paths in the responses will have the `prod_env/cache` prefix baked in which means that if the user sets the `--local-path` flag to something else, this breaks all paths in the responses cache and the `scenario_state.json` files, and the user has to do some manual migration.
3. It's not really clear what relative file paths should be relative to - this pull request takes the position that it should be relative to the current working directory, but there are advantages to making it relative to the `prod_env/cache` , or to the `benchmark_output`.
4. We do not upload anything in `prod_env/cache`, which means that the publicly uploaded `scenario_state.json` files will contain paths to unavailable files.
5. Even if I wanted to upload the files in `prod_env/cache` to a web server, I would be forced to upload them with `prod_env/cache` as part of the URL so that the links keep working.
6. Images from different run specs and different run suites are mixed, so I can't upload the images for a specific suite or run - I'd have to write a script to do this.
7. 
Some possible alternatives:

1. Remove FileCache: Store the image in the response cache. Both SQLite and MongoDB have binary blob types that do this (or we can base64 encode, which is less efficient).
2. Write out the files to the run folder. i.e. `/benchmark_output/run/some_scenario:model=some_model/media/77437ea482144bf7b9275a0acee997db.png` In `MediaObject` in `RequestResponse` in the `ScenarioState`, the paths will be `media/77437ea482144bf7b9275a0acee997db.png` - relative to the location of `scenario_state.json`.

Summarizing some stuff from a Slack converation:

@teetone:

> I thought of another solution that is decent, quick to implement and addresses most of the issues you mentioned. We can cache the images in prod_env when the clients receive a request. When we generate the Scenario State, we can create a copy of the files in benchmark_output and specify the copied path in Scenario State, so benchmark_output will have the copy of the images. The downside is we now will have an extra copy of the images.

Me:

> That sounds reasonable as well - a second copy is fine because the cache copy is temporary, so we only have one permanent copy.
> It wouldn't address the baked-in paths in the response cache, but I would be okay with compromising on that.
> Also, with this, we wouldn't need the base64 encoding any more - we can just link to the actual file. So we actually save on storage by eliminating the base64 encoded copy (because binary is more efficient than base64)

Percy:

> Having a temporary cache in `prod_env/cache` makes sense because it mirrors the way we cache everything else.
> But at that point, why not just store in mongo or the sqlite db?

> To understand the proposal for `scenario_state.json`, the files would just be parallel to that?  It seems like `prod_env/cache` should not appear in the paths.

teetone commented on src/helm/benchmark/window_services/image_generation/clip_window_service.py:
```diff
@@ -0,0 +1,43 @@
+    def truncate_from_right(self, text: str, expected_completion_token_length: int = 0) -> str:
+        result: str = self.decode(self.encode(text, truncation=True, max_length=self.max_request_length).tokens)
+
+        # HACK: For the vast majority of cases, the above logic works, but there are a few where the
+        # token count exceeds `max_length` by 1.
+        while not self.fits_within_context_window(result):
+            result = result[:-1]
```
Removing one character may or may not truncate by one token, but since the lengths of tokens can vary, removing character by character will eventually remove tokens one by one.
